File: gallery/invaders/invaders.swift
import Foundation

var r_key_channel: DispatchQueue? = nil
var r_key_queue: [String] = []
var r_key_queue_lock = NSLock()

#if os(Windows)
@_silgen_name("_kbhit") func _kbhit() -> Int32
@_silgen_name("_getch") func _getch() -> Int32

func r_setup_raw_mode() {
    // No setup needed on Windows
}

func r_read_key() -> String? {
    if _kbhit() == 0 { return nil }
    let ch = _getch()
    if ch == 3 { // Ctrl+C
        print("\u{1b}[?25h", terminator: "")
        exit(0)
    }
    if ch == 224 || ch == 0 { // Extended key
        let ch2 = _getch()
        switch ch2 {
        case 72: return "up"
        case 80: return "down"
        case 75: return "left"
        case 77: return "right"
        default: return ""
        }
    }
    switch ch {
    case 8: return "backspace"
    case 9: return "tab"
    case 13: return "enter"
    case 27: return "escape"
    case 32: return "space"
    default: return String(UnicodeScalar(UInt8(ch)))
    }
}
#else
func r_setup_raw_mode() {
    let _ = system("stty cbreak min 1 -echo")
}

func r_read_key() -> String? {
    var buf = [UInt8](repeating: 0, count: 3)
    let n = read(STDIN_FILENO, &buf, 1)
    if n <= 0 { return nil }
    if buf[0] == 3 { // Ctrl+C
        print("\u{1b}[?25h", terminator: "")
        exit(0)
    }
    if buf[0] == 27 { // Escape sequence
        let _ = read(STDIN_FILENO, &buf[1], 2)
        if buf[1] == 91 {
            switch buf[2] {
            case 65: return "up"
            case 66: return "down"
            case 67: return "right"
            case 68: return "left"
            default: break
            }
        }
        return "escape"
    }
    switch buf[0] {
    case 8, 127: return "backspace"
    case 9: return "tab"
    case 10, 13: return "enter"
    case 32: return "space"
    default: return String(UnicodeScalar(buf[0]))
    }
}
#endif

func ==(l: Alien, r: Alien) -> Bool {
  return l === r
}
class Alien : Equatable  { 
  var x : Int = 0
  var y : Int = 0
  var prevX : Int = 0
  var prevY : Int = 0
  var alive : Bool = true
  var wasAlive : Bool = true
  init(startX : Int, startY : Int ) {
    self.x = startX;
    self.y = startY;
    self.prevX = startX;
    self.prevY = startY;
  }
  func savePrev() -> Void {
    self.prevX = self.x;
    self.prevY = self.y;
    self.wasAlive = self.alive;
  }
}
func ==(l: Bullet, r: Bullet) -> Bool {
  return l === r
}
class Bullet : Equatable  { 
  var x : Int = 0
  var y : Int = 0
  var prevX : Int = 0
  var prevY : Int = 0
  var active : Bool = true
  var wasActive : Bool = true
  init(startX : Int, startY : Int ) {
    self.x = startX;
    self.y = startY;
    self.prevX = startX;
    self.prevY = startY;
  }
  func savePrev() -> Void {
    self.prevX = self.x;
    self.prevY = self.y;
    self.wasActive = self.active;
  }
}
func ==(l: Invaders, r: Invaders) -> Bool {
  return l === r
}
class Invaders : Equatable  { 
  var WIDTH : Int = 40
  var HEIGHT : Int = 18
  var PLAYER_Y : Int = 16
  var playerX : Int = 20
  var drawnPlayerX : Int = -1
  var aliens : [Alien] = [Alien]()
  var bullets : [Bullet] = [Bullet]()
  var score : Int = 0
  var prevScore : Int = 0
  var gameOver : Bool = false
  var gameWon : Bool = false
  var alienDirection : Int = 1
  var frameCount : Int = 0
  var alienMoveDelay : Int = 8
  var firstRender : Bool = true
  init( ) {
    self.initAliens()
  }
  func initAliens() -> Void {
    var row : Int = 0
    while (row < 3) {
      var col : Int = 0
      while (col < 8) {
        let ax : Int = (col * 4) + 4
        let ay : Int = row + 2
        let alien : Alien = Alien(startX : ax, startY : ay)
        self.aliens.append(alien)
        col = col + 1;
      }
      row = row + 1;
    }
  }
  func drawAt(x : Int, y : Int, ch : String) -> Void {
    print("\u{1b}[\(y + 1);\(x + 1)H", terminator: "")
    print(ch, terminator: "")
  }
  func eraseAt(x : Int, y : Int) -> Void {
    print("\u{1b}[\(y + 1);\(x + 1)H", terminator: "")
    print(" ", terminator: "")
  }
  func drawBorders() -> Void {
    print("\u{1b}[\(1);\(1)H", terminator: "")
    var i : Int = 0
    let borderW : Int = self.WIDTH + 2
    while (i < borderW) {
      print("=", terminator: "")
      i = i + 1;
    }
    var y : Int = 0
    while (y < self.HEIGHT) {
      print("\u{1b}[\(y + 2);\(1)H", terminator: "")
      print("|", terminator: "")
      let rightX : Int = self.WIDTH + 2
      print("\u{1b}[\(y + 2);\(rightX)H", terminator: "")
      print("|", terminator: "")
      y = y + 1;
    }
    let bottomY : Int = self.HEIGHT + 2
    print("\u{1b}[\(bottomY);\(1)H", terminator: "")
    var j : Int = 0
    while (j < borderW) {
      print("=", terminator: "")
      j = j + 1;
    }
    let scoreY : Int = self.HEIGHT + 3
    print("\u{1b}[\(scoreY);\(1)H", terminator: "")
    print("Score:       |  Arrows=move  SPACE=shoot  Q=quit", terminator: "")
  }
  func render() -> Void {
    if ( self.firstRender ) {
      print("\u{1b}[2J\u{1b}[H", terminator: "")
      self.drawBorders()
      self.firstRender = false;
    }
    if ( self.drawnPlayerX >= 0 ) {
      if ( self.drawnPlayerX != self.playerX ) {
        self.eraseAt(x : self.drawnPlayerX, y : self.PLAYER_Y)
      }
    }
    self.drawAt(x : self.playerX, y : self.PLAYER_Y, ch : "A")
    self.drawnPlayerX = self.playerX;
    for (idx, alien) in self.aliens.enumerated() {
      if ( alien.wasAlive ) {
        var moved : Bool = false
        if ( alien.prevX != alien.x ) {
          moved = true;
        }
        if ( alien.prevY != alien.y ) {
          moved = true;
        }
        if ( moved ) {
          let oldAx : Int = alien.prevX + 1
          self.eraseAt(x : oldAx, y : alien.prevY)
        }
        if ( alien.alive == false ) {
          let killAx : Int = alien.x + 1
          self.eraseAt(x : killAx, y : alien.y)
        }
      }
      if ( alien.alive ) {
        let drawX : Int = alien.x + 1
        self.drawAt(x : drawX, y : alien.y, ch : "W")
      }
    }
    for (idx2, bullet) in self.bullets.enumerated() {
      if ( bullet.wasActive ) {
        let oldBx : Int = bullet.prevX + 1
        self.eraseAt(x : oldBx, y : bullet.prevY)
      }
      if ( bullet.active ) {
        let bx : Int = bullet.x + 1
        self.drawAt(x : bx, y : bullet.y, ch : "|")
      }
    }
    if ( self.score != self.prevScore ) {
      let scoreY : Int = self.HEIGHT + 3
      print("\u{1b}[\(scoreY);\(8)H", terminator: "")
      let scoreStr : String = String(self.score) + "   "
      print(scoreStr, terminator: "")
    }
    let endY : Int = self.HEIGHT + 4
    print("\u{1b}[\(endY);\(1)H", terminator: "")
  }
  func savePrevState() -> Void {
    self.prevScore = self.score;
    for (idx, alien) in self.aliens.enumerated() {
      alien.savePrev()
    }
    for (idx2, bullet) in self.bullets.enumerated() {
      bullet.savePrev()
    }
  }
  func shoot() -> Void {
    let bulletY : Int = self.PLAYER_Y - 1
    let bullet : Bullet = Bullet(startX : self.playerX, startY : bulletY)
    self.bullets.append(bullet)
  }
  func moveLeft() -> Void {
    if ( self.playerX > 2 ) {
      self.playerX = self.playerX - 1;
    }
  }
  func moveRight() -> Void {
    let maxX : Int = self.WIDTH - 1
    if ( self.playerX < maxX ) {
      self.playerX = self.playerX + 1;
    }
  }
  func updateBullets() -> Void {
    for (idx, bullet) in self.bullets.enumerated() {
      if ( bullet.active ) {
        bullet.y = bullet.y - 1;
        if ( bullet.y < 1 ) {
          bullet.active = false;
        }
      }
    }
  }
  func countAlive() -> Int {
    var count : Int = 0
    for (idx, a) in self.aliens.enumerated() {
      if ( a.alive ) {
        count = count + 1;
      }
    }
    return count;
  }
  func updateAliens() -> Void {
    let moveFrame : Int = self.frameCount % self.alienMoveDelay
    if ( moveFrame != 0 ) {
      return;
    }
    var shouldMoveDown : Bool = false
    var minX : Int = 999
    var maxX : Int = 0
    for (idx, alien) in self.aliens.enumerated() {
      if ( alien.alive ) {
        if ( alien.x < minX ) {
          minX = alien.x;
        }
        if ( alien.x > maxX ) {
          maxX = alien.x;
        }
      }
    }
    let rightBound : Int = self.WIDTH - 2
    if ( self.alienDirection > 0 ) {
      if ( maxX >= rightBound ) {
        self.alienDirection = -1;
        shouldMoveDown = true;
      }
    } else {
      if ( minX <= 1 ) {
        self.alienDirection = 1;
        shouldMoveDown = true;
      }
    }
    for (idx2, alien_1) in self.aliens.enumerated() {
      if ( alien_1.alive ) {
        alien_1.x = alien_1.x + self.alienDirection;
        if ( shouldMoveDown ) {
          alien_1.y = alien_1.y + 1;
          if ( alien_1.y >= self.PLAYER_Y ) {
            self.gameOver = true;
          }
        }
      }
    }
    let aliveCount : Int = self.countAlive()
    if ( aliveCount < 12 ) {
      self.alienMoveDelay = 5;
    }
    if ( aliveCount < 6 ) {
      self.alienMoveDelay = 3;
    }
    if ( aliveCount < 3 ) {
      self.alienMoveDelay = 1;
    }
  }
  func checkCollisions() -> Void {
    for (bidx, bullet) in self.bullets.enumerated() {
      if ( bullet.active ) {
        for (aidx, alien) in self.aliens.enumerated() {
          if ( alien.alive ) {
            if ( bullet.x == alien.x ) {
              if ( bullet.y == alien.y ) {
                bullet.active = false;
                alien.alive = false;
                self.score = self.score + 10;
              }
            }
          }
        }
      }
    }
  }
  func checkWin() -> Void {
    let aliveCount : Int = self.countAlive()
    if ( aliveCount == 0 ) {
      self.gameWon = true;
      self.gameOver = true;
    }
  }
  func update() -> Void {
    self.frameCount = self.frameCount + 1;
    self.updateBullets()
    self.updateAliens()
    self.checkCollisions()
    self.checkWin()
  }
  func endGame() -> Void {
    print("\u{1b}[2J\u{1b}[H", terminator: "")
    print("\u{1b}[\(1);\(1)H", terminator: "")
    if ( self.gameWon ) {
      print("=== YOU WIN! ===")
    } else {
      print("=== GAME OVER ===")
    }
    let finalMsg : String = "Final Score: " + String(self.score)
    print(finalMsg)
    print("\u{1b}[?25h", terminator: "")
  }
  func gameLoop() -> Void {
    while (self.gameOver == false) {
      self.savePrevState()
      self.update()
      self.render()
      let key : String = { () -> String in r_key_queue_lock.lock(); defer { r_key_queue_lock.unlock() }; if r_key_queue.isEmpty { return "" }; return r_key_queue.removeFirst() }()
      if ( key != "" ) {
        self.handleKey(key : key)
      }
      Thread.sleep(forTimeInterval: Double(50) / 1000.0)
    }
    self.endGame()
  }
  func handleKey(key : String) -> Void {
    if ( key == "left" ) {
      self.moveLeft()
    }
    if ( key == "right" ) {
      self.moveRight()
    }
    if ( key == "a" ) {
      self.moveLeft()
    }
    if ( key == "A" ) {
      self.moveLeft()
    }
    if ( key == "d" ) {
      self.moveRight()
    }
    if ( key == "D" ) {
      self.moveRight()
    }
    if ( key == " " ) {
      self.shoot()
    }
    if ( key == "space" ) {
      self.shoot()
    }
    if ( key == "q" ) {
      self.gameOver = true;
    }
    if ( key == "Q" ) {
      self.gameOver = true;
    }
  }
}
// Swift 6 entry point
@main
struct Main {
  static func main() {
    let game : Invaders = Invaders()
    print("=== SPACE INVADERS ===")
    print("")
    print("Controls:")
    print("  LEFT/RIGHT - Move")
    print("  SPACE      - Shoot")
    print("  Q          - Quit")
    print("")
    print("Starting game...")
    let key : String = ""
    _ = key
    r_key_channel = DispatchQueue(label: "keypress")
    r_key_channel!.async {
      r_setup_raw_mode()
      while true {
        if let k = r_read_key(), !k.isEmpty {
          r_key_queue_lock.lock()
          r_key_queue.append(k)
          r_key_queue_lock.unlock()
        }
        Thread.sleep(forTimeInterval: 0.01)
      }
    }
    print("\u{1b}[?25l", terminator: "")
    game.gameLoop()
  }
}

File: gallery/invaders/invaders.cpp
#include  <memory>
#include  "variant.hpp"
#include  <vector>
#include  <string>
#include  <iostream>
#include  <chrono>
#include  <thread>
#include  <mutex>
#include  <queue>

// define classes here to avoid compiler errors
class Alien;
class Bullet;
class Invaders;

typedef mpark::variant<std::shared_ptr<Alien>, std::shared_ptr<Bullet>, std::shared_ptr<Invaders>, int, std::string, bool, double>  r_union_Any;

#include <string>
#include <queue>
#include <mutex>

std::queue<std::string> r_key_queue;
std::mutex r_key_mutex;

#ifdef _WIN32
#include <conio.h>
void r_setup_raw_mode() {}
std::string r_read_key() {
    if (!_kbhit()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
        return "";
    }
    int ch = _getch();
    if (ch == 3) { std::cout << "\x1b[?25h"; exit(0); }
    if (ch == 224 || ch == 0) {
        int ch2 = _getch();
        switch (ch2) {
            case 72: return "up";
            case 80: return "down";
            case 75: return "left";
            case 77: return "right";
        }
        return "";
    }
    switch (ch) {
        case 8: return "backspace";
        case 9: return "tab";
        case 13: return "enter";
        case 27: return "escape";
        case 32: return "space";
    }
    return std::string(1, (char)ch);
}
#else
#include <termios.h>
#include <unistd.h>
void r_setup_raw_mode() {
    system("stty cbreak min 1 -echo");
}
std::string r_read_key() {
    char buf[3];
    if (read(STDIN_FILENO, buf, 1) <= 0) return "";
    if (buf[0] == 3) { std::cout << "\x1b[?25h"; exit(0); }
    if (buf[0] == 27) {
        read(STDIN_FILENO, buf+1, 2);
        if (buf[1] == 91) {
            switch (buf[2]) {
                case 65: return "up";
                case 66: return "down";
                case 67: return "right";
                case 68: return "left";
            }
        }
        return "escape";
    }
    switch (buf[0]) {
        case 8: case 127: return "backspace";
        case 9: return "tab";
        case 10: case 13: return "enter";
        case 32: return "space";
    }
    return std::string(1, buf[0]);
}
#endif


// header definitions
class Alien : public std::enable_shared_from_this<Alien>  { 
  public :
    int x;
    int y;
    int prevX;
    int prevY;
    bool alive;
    bool wasAlive;
    /* class constructor */ 
    Alien( int startX , int startY  );
    /* instance methods */ 
    void savePrev();
};
class Bullet : public std::enable_shared_from_this<Bullet>  { 
  public :
    int x;
    int y;
    int prevX;
    int prevY;
    bool active;
    bool wasActive;
    /* class constructor */ 
    Bullet( int startX , int startY  );
    /* instance methods */ 
    void savePrev();
};
class Invaders : public std::enable_shared_from_this<Invaders>  { 
  public :
    int WIDTH;
    int HEIGHT;
    int PLAYER_Y;
    int playerX;
    int drawnPlayerX;
    std::vector<std::shared_ptr<Alien>> aliens;
    std::vector<std::shared_ptr<Bullet>> bullets;
    int score;
    int prevScore;
    bool gameOver;
    bool gameWon;
    int alienDirection;
    int frameCount;
    int alienMoveDelay;
    bool firstRender;
    /* class constructor */ 
    Invaders( );
    /* static methods */ 
    static void m();
    /* instance methods */ 
    void initAliens();
    void drawAt( int x , int y , std::string ch );
    void eraseAt( int x , int y );
    void drawBorders();
    void render();
    void savePrevState();
    void shoot();
    void moveLeft();
    void moveRight();
    void updateBullets();
    int countAlive();
    void updateAliens();
    void checkCollisions();
    void checkWin();
    void update();
    void endGame();
    void gameLoop();
    void handleKey( std::string key );
};

int __g_argc;
char **__g_argv;
Alien::Alien( int startX , int startY  ) {
  this->x = 0;
  this->y = 0;
  this->prevX = 0;
  this->prevY = 0;
  this->alive = true;
  this->wasAlive = true;
  x = startX;
  y = startY;
  prevX = startX;
  prevY = startY;
}
void  Alien::savePrev() {
  prevX = x;
  prevY = y;
  wasAlive = alive;
}
Bullet::Bullet( int startX , int startY  ) {
  this->x = 0;
  this->y = 0;
  this->prevX = 0;
  this->prevY = 0;
  this->active = true;
  this->wasActive = true;
  x = startX;
  y = startY;
  prevX = startX;
  prevY = startY;
}
void  Bullet::savePrev() {
  prevX = x;
  prevY = y;
  wasActive = active;
}
Invaders::Invaders( ) {
  this->WIDTH = 40;
  this->HEIGHT = 18;
  this->PLAYER_Y = 16;
  this->playerX = 20;
  this->drawnPlayerX = -1;
  this->score = 0;
  this->prevScore = 0;
  this->gameOver = false;
  this->gameWon = false;
  this->alienDirection = 1;
  this->frameCount = 0;
  this->alienMoveDelay = 8;
  this->firstRender = true;
  this->initAliens();
}
void  Invaders::initAliens() {
  int row = 0;
  while (row < 3) {
    int col = 0;
    while (col < 8) {
      int ax = (col * 4) + 4;
      int ay = row + 2;
      std::shared_ptr<Alien> alien =  std::make_shared<Alien>(ax, ay);
      aliens.push_back( alien  );
      col = col + 1;
    }
    row = row + 1;
  }
}
void  Invaders::drawAt( int x , int y , std::string ch ) {
  std::cout << "\x1b[" << y + 1 << ";" << x + 1 << "H" << std::flush;
  std::cout << ch << std::flush;
}
void  Invaders::eraseAt( int x , int y ) {
  std::cout << "\x1b[" << y + 1 << ";" << x + 1 << "H" << std::flush;
  std::cout << std::string(" ") << std::flush;
}
void  Invaders::drawBorders() {
  std::cout << "\x1b[" << 1 << ";" << 1 << "H" << std::flush;
  int i = 0;
  int borderW = WIDTH + 2;
  while (i < borderW) {
    std::cout << std::string("=") << std::flush;
    i = i + 1;
  }
  int y = 0;
  while (y < HEIGHT) {
    std::cout << "\x1b[" << y + 2 << ";" << 1 << "H" << std::flush;
    std::cout << std::string("|") << std::flush;
    int rightX = WIDTH + 2;
    std::cout << "\x1b[" << y + 2 << ";" << rightX << "H" << std::flush;
    std::cout << std::string("|") << std::flush;
    y = y + 1;
  }
  int bottomY = HEIGHT + 2;
  std::cout << "\x1b[" << bottomY << ";" << 1 << "H" << std::flush;
  int j = 0;
  while (j < borderW) {
    std::cout << std::string("=") << std::flush;
    j = j + 1;
  }
  int scoreY = HEIGHT + 3;
  std::cout << "\x1b[" << scoreY << ";" << 1 << "H" << std::flush;
  std::cout << std::string("Score:       |  Arrows=move  SPACE=shoot  Q=quit") << std::flush;
}
void  Invaders::render() {
  if ( firstRender ) {
    std::cout << "\x1b[2J\x1b[H" << std::flush;
    this->drawBorders();
    firstRender = false;
  }
  if ( drawnPlayerX >= 0 ) {
    if ( drawnPlayerX != playerX ) {
      this->eraseAt(drawnPlayerX, PLAYER_Y);
    }
  }
  this->drawAt(playerX, PLAYER_Y, std::string("A"));
  drawnPlayerX = playerX;
  for ( int idx = 0; idx != (int)(aliens.size()); idx++) {
    std::shared_ptr<Alien> alien = aliens.at(idx);
    if ( alien->wasAlive ) {
      bool moved = false;
      if ( alien->prevX != alien->x ) {
        moved = true;
      }
      if ( alien->prevY != alien->y ) {
        moved = true;
      }
      if ( moved ) {
        int oldAx = alien->prevX + 1;
        this->eraseAt(oldAx, alien->prevY);
      }
      if ( alien->alive == false ) {
        int killAx = alien->x + 1;
        this->eraseAt(killAx, alien->y);
      }
    }
    if ( alien->alive ) {
      int drawX = alien->x + 1;
      this->drawAt(drawX, alien->y, std::string("W"));
    }
  }
  for ( int idx2 = 0; idx2 != (int)(bullets.size()); idx2++) {
    std::shared_ptr<Bullet> bullet = bullets.at(idx2);
    if ( bullet->wasActive ) {
      int oldBx = bullet->prevX + 1;
      this->eraseAt(oldBx, bullet->prevY);
    }
    if ( bullet->active ) {
      int bx = bullet->x + 1;
      this->drawAt(bx, bullet->y, std::string("|"));
    }
  }
  if ( score != prevScore ) {
    int scoreY = HEIGHT + 3;
    std::cout << "\x1b[" << scoreY << ";" << 8 << "H" << std::flush;
    std::string scoreStr = std::to_string(score) + std::string("   ");
    std::cout << scoreStr << std::flush;
  }
  int endY = HEIGHT + 4;
  std::cout << "\x1b[" << endY << ";" << 1 << "H" << std::flush;
}
void  Invaders::savePrevState() {
  prevScore = score;
  for ( int idx = 0; idx != (int)(aliens.size()); idx++) {
    std::shared_ptr<Alien> alien = aliens.at(idx);
    alien->savePrev();
  }
  for ( int idx2 = 0; idx2 != (int)(bullets.size()); idx2++) {
    std::shared_ptr<Bullet> bullet = bullets.at(idx2);
    bullet->savePrev();
  }
}
void  Invaders::shoot() {
  int bulletY = PLAYER_Y - 1;
  std::shared_ptr<Bullet> bullet =  std::make_shared<Bullet>(playerX, bulletY);
  bullets.push_back( bullet  );
}
void  Invaders::moveLeft() {
  if ( playerX > 2 ) {
    playerX = playerX - 1;
  }
}
void  Invaders::moveRight() {
  int maxX = WIDTH - 1;
  if ( playerX < maxX ) {
    playerX = playerX + 1;
  }
}
void  Invaders::updateBullets() {
  for ( int idx = 0; idx != (int)(bullets.size()); idx++) {
    std::shared_ptr<Bullet> bullet = bullets.at(idx);
    if ( bullet->active ) {
      bullet->y = bullet->y - 1;
      if ( bullet->y < 1 ) {
        bullet->active = false;
      }
    }
  }
}
int  Invaders::countAlive() {
  int count = 0;
  for ( int idx = 0; idx != (int)(aliens.size()); idx++) {
    std::shared_ptr<Alien> a = aliens.at(idx);
    if ( a->alive ) {
      count = count + 1;
    }
  }
  return count;
}
void  Invaders::updateAliens() {
  int moveFrame = frameCount % alienMoveDelay;
  if ( moveFrame != 0 ) {
    return;
  }
  bool shouldMoveDown = false;
  int minX = 999;
  int maxX = 0;
  for ( int idx = 0; idx != (int)(aliens.size()); idx++) {
    std::shared_ptr<Alien> alien = aliens.at(idx);
    if ( alien->alive ) {
      if ( alien->x < minX ) {
        minX = alien->x;
      }
      if ( alien->x > maxX ) {
        maxX = alien->x;
      }
    }
  }
  int rightBound = WIDTH - 2;
  if ( alienDirection > 0 ) {
    if ( maxX >= rightBound ) {
      alienDirection = -1;
      shouldMoveDown = true;
    }
  } else {
    if ( minX <= 1 ) {
      alienDirection = 1;
      shouldMoveDown = true;
    }
  }
  for ( int idx2 = 0; idx2 != (int)(aliens.size()); idx2++) {
    std::shared_ptr<Alien> alien_1 = aliens.at(idx2);
    if ( alien_1->alive ) {
      alien_1->x = alien_1->x + alienDirection;
      if ( shouldMoveDown ) {
        alien_1->y = alien_1->y + 1;
        if ( alien_1->y >= PLAYER_Y ) {
          gameOver = true;
        }
      }
    }
  }
  int aliveCount = this->countAlive();
  if ( aliveCount < 12 ) {
    alienMoveDelay = 5;
  }
  if ( aliveCount < 6 ) {
    alienMoveDelay = 3;
  }
  if ( aliveCount < 3 ) {
    alienMoveDelay = 1;
  }
}
void  Invaders::checkCollisions() {
  for ( int bidx = 0; bidx != (int)(bullets.size()); bidx++) {
    std::shared_ptr<Bullet> bullet = bullets.at(bidx);
    if ( bullet->active ) {
      for ( int aidx = 0; aidx != (int)(aliens.size()); aidx++) {
        std::shared_ptr<Alien> alien = aliens.at(aidx);
        if ( alien->alive ) {
          if ( bullet->x == alien->x ) {
            if ( bullet->y == alien->y ) {
              bullet->active = false;
              alien->alive = false;
              score = score + 10;
            }
          }
        }
      }
    }
  }
}
void  Invaders::checkWin() {
  int aliveCount = this->countAlive();
  if ( aliveCount == 0 ) {
    gameWon = true;
    gameOver = true;
  }
}
void  Invaders::update() {
  frameCount = frameCount + 1;
  this->updateBullets();
  this->updateAliens();
  this->checkCollisions();
  this->checkWin();
}
void  Invaders::endGame() {
  std::cout << "\x1b[2J\x1b[H" << std::flush;
  std::cout << "\x1b[" << 1 << ";" << 1 << "H" << std::flush;
  if ( gameWon ) {
    std::cout << std::string("=== YOU WIN! ===") << std::endl;
  } else {
    std::cout << std::string("=== GAME OVER ===") << std::endl;
  }
  std::string finalMsg = std::string("Final Score: ") + std::to_string(score);
  std::cout << finalMsg << std::endl;
  std::cout << "\x1b[?25h" << std::flush;
}
void  Invaders::gameLoop() {
  while (gameOver == false) {
    this->savePrevState();
    this->update();
    this->render();
    std::string key = [&]() -> std::string { std::lock_guard<std::mutex> lock(r_key_mutex); if (r_key_queue.empty()) return ""; std::string k = r_key_queue.front(); r_key_queue.pop(); return k; }();
    if ( key != std::string("") ) {
      this->handleKey(key);
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
  }
  this->endGame();
}
void  Invaders::handleKey( std::string key ) {
  if ( key == std::string("left") ) {
    this->moveLeft();
  }
  if ( key == std::string("right") ) {
    this->moveRight();
  }
  if ( key == std::string("a") ) {
    this->moveLeft();
  }
  if ( key == std::string("A") ) {
    this->moveLeft();
  }
  if ( key == std::string("d") ) {
    this->moveRight();
  }
  if ( key == std::string("D") ) {
    this->moveRight();
  }
  if ( key == std::string(" ") ) {
    this->shoot();
  }
  if ( key == std::string("space") ) {
    this->shoot();
  }
  if ( key == std::string("q") ) {
    gameOver = true;
  }
  if ( key == std::string("Q") ) {
    gameOver = true;
  }
}
int main(int argc, char* argv[]) {
  __g_argc = argc;
  __g_argv = argv;
  std::shared_ptr<Invaders> game =  std::make_shared<Invaders>();
  std::cout << std::string("=== SPACE INVADERS ===") << std::endl;
  std::cout << std::string("") << std::endl;
  std::cout << std::string("Controls:") << std::endl;
  std::cout << std::string("  LEFT/RIGHT - Move") << std::endl;
  std::cout << std::string("  SPACE      - Shoot") << std::endl;
  std::cout << std::string("  Q          - Quit") << std::endl;
  std::cout << std::string("") << std::endl;
  std::cout << std::string("Starting game...") << std::endl;
  std::string key = std::string("");
  std::thread([&]() {
    r_setup_raw_mode();
    while (true) {
      std::string k = r_read_key();
      if (!k.empty()) {
        key = k;
        {
          std::lock_guard<std::mutex> lock(r_key_mutex);
          r_key_queue.push(k);
        }
        game->handleKey(key);
      }
    }
  }).detach();
  std::cout << "\x1b[?25l" << std::flush;
  game->gameLoop();
  return 0;
}
